use pwsh and add work-dir option

gstrauss opened the pull request:
* use pwsh (instead of powershell)
* add work-dir option (e.g. to allow specifying 'D:' instead of default 'C:')

I find powershell to be fugly and touch it with a ten-foot-pole when I have to.
Suggestions for improvements to my syntax are welcome.  Thanks.

gstrauss committed:
use pwsh instead of powershell

shell: pwsh        # uses Powershell 7 (latest)
shell: powershell  # uses Powershell 5 (stuck at Powershell 5)

Signed-off-by: Glenn Strauss <gstrauss@gluelogic.com>

gstrauss committed:
use ternary operator with pwsh 7

Signed-off-by: Glenn Strauss <gstrauss@gluelogic.com>

gstrauss commented:
@jon-turney any suggestions/pointers why CI is failing here to find `sponge.exe`?  or why the multiple sites and extra pubkeys test is failing for me?

gstrauss commented:
Wow.  The change from powershell and pwsh is causing the tests to fail.  (expletive deleted)

gstrauss commented:
I had to apply this patch to the PR to change the primary instance of 'pwsh' back to 'powershell', or else 3 of the 12 CI tests failed.
```
--- a/action.yml
+++ b/action.yml
@@ -144,7 +144,7 @@ runs:
         # because setup is a Windows GUI app, make it part of a pipeline to make
         # PowerShell wait for it to exit
         & $setupExe $args | Out-Default
-      shell: pwsh
+      shell: powershell
 
     - if: ${{ inputs.add-to-path  == 'true' }}
       run: |
```

I could use some help from someone more familiar with powershell and pwsh to see if the tests need to be updated or if something else is incompatible between powershell (5) and pwsh (7)

gstrauss commented:
@jon-turney some quick (non-scientific n=2) tests running the lighttpd CI twice on C: took 4 min 12 sec, and 5 min 10 sec; and twice on D: took 2 min 24 sec, and 1 min 47 sec.  While there is some slippage comparing times when then package download time from mirrors.kernel.org can vary, in my single-run test the C: drive took about 1 min 10 sec, and the D: drive took a few seconds less than that, so the rest of the time difference is likely the Windows bloatware "protecting" the C:\ drive (which contains C:\Windows) while setup.exe is extracting and installing the packages to the C: drive versus the D: drive.

Running the same CI (once) were C:\setup.exe and C:\cygwin-packages, but having the install-dir D:\cygwin took 2 min 25 sec, so the existing install-dir option in the workflow appears to get most of the benefit to not being on C:\.  While using the D:\ drive for D:\setup.exe and D:\cygwin-packages is not worse than using C:\ for those as long as the installation directory is on D:\cygwin, I would understand if you think the install-dir option is sufficient for specifying D:\cygwin rather than the new work-vol option in this PR.

In the lighttpd CI, I am using `pwsh` instead of `powershell`.  I still wonder why some of the cygwin tests fail under `pwsh`.

gstrauss commented:
Turns out that powershell and pwsh did a fugly reimplementation of awk more than 40 years later after awk.

You're more than welcome to suggest prettier syntax.  The problem was that some of the *basic string manipulation syntax* changed between powershell (5) and pwsh (7) and the input strings were no longer being split under pwsh as they were under powershell.

gstrauss commented:
@jon-turney Ready for review.  Please review the two commits separately.

gstrauss commented:
@jon-turney given the extreme performance difference between D:\ and C:\ for `install-dir`, what would you think about making the default `install-dir` `D:\cygwin` and, when `install-dir` is not specified, creating a symlink from `C:\cygwin` -> `D:\cygwin` for compatibility with existing CI .yml which use the default `install-dir` (`C:\cygwin`) ?

jon-turney commented on action.yml:
```diff
@@ -51,9 +55,11 @@ runs:
           echo "unknown platform $platform"
           exit 1
         }
+        $vol = '${{ inputs.work-vol }}'
```
Perhaps some bullet-proofing here?

If it's one character, append a ':'?
If it's exactly two characters it must end with a ':'
I guess theoretically this could be a UNC or long path name, so otherwise it must start '\\'

Or just an error if it's not of the expected form?

gstrauss replied:
If it is a bad path prefix, download of setup.exe will fail soon after.  Is that not sufficient?

BTW, there are no such bullet-proofing checks on `install-dir`.

jon-turney replied:
Yeah, this is probably excessive.

Maybe if someone mistakenly uses 'C' rather than 'C:' you get a directory named C in cwd, or something, but I guess they'll learn not to do that 😖

jon-turney commented on action.yml:
```diff
@@ -124,16 +132,19 @@ runs:
         # run login shell to copy skeleton profile files
-        ${{ inputs.install-dir }}\bin\bash.exe --login
+        $installDir = '${{ inputs.install-dir }}' -ne '' ? '${{ inputs.install-dir }}' : '${{ inputs.work-vol }}\cygwin'
+        & "$installDir\bin\bash.exe" --login
```
I think this could just be 'bash --login', since we've just prepended the install-dir to the PATH in the previous step?

gstrauss replied:
The previous step runs only `- if: ${{ inputs.add-to-path  == 'true' }}`, which is the default, but can be disabled.

jon-turney commented:
> The problem was that some of the _basic string manipulation syntax_ changed between powershell (5) and pwsh (7) and the input strings were no longer being split under pwsh as they were under powershell.

There's no way that's true.... (does some web searches)... it's true :facepalm:

jon-turney commented:
> Please review the two commits separately.

Made a few minor comments of the packages dir.

I don't have any comments on the "use pwsh" commit (my eyes glaze over at it's crufty syntax as well).  But why is it here? Does the other commit depend on it in some way?

jon-turney commented:
Thanks for working on this!

So, my concerns with all this are:

(i) is there some reasonable indication that D:\ will exist and be faster on the VMs we run on?
(ii) Is there enough a guarantee about how much free space there is on D:\ ?

I found [1], which seems to answer (i). 

[1] https://learn.microsoft.com/en-us/azure/virtual-machines/managed-disks-overview#temporary-disk

> given the extreme performance difference between D:\ and C:\ for `install-dir`, what would you think about making the default `install-dir` `D:\cygwin` and, when `install-dir` is not specified, creating a symlink from `C:\cygwin` -> `D:\cygwin` for compatibility with existing CI .yml which use the default `install-dir` (`C:\cygwin`) ?

Yeah, D:\ being the default seems like the sensible thing to do.

gstrauss commented:
> I don't have any comments on the "use pwsh" commit (my eyes glaze over at it's crufty syntax as well). But why is it here? Does the other commit depend on it in some way?

The commit to switch from `powershell` to `pwsh` started with experiments in #25.  Still, I think it is a good idea to move to the "better-supported" version of powershell from Microsoft, where "better" and "supported" are both highly up to interpretation because this is after all a Microsoft product.

gstrauss commented:
`Get-Volume` run from the github runner returns:
```
DriveLetter FriendlyName      FileSystemType DriveType HealthStatus OperationalStatus SizeRemaining      Size
----------- ------------      -------------- --------- ------------ ----------------- -------------      ----
D           Temporary Storage NTFS           Fixed     Healthy      OK                    147.02 GB    150 GB
A                             Unknown        Removable Healthy      Unknown                     0 B       0 B
            System Reserved   NTFS           Fixed     Healthy      OK                    464.96 MB    500 MB
C           Windows           NTFS           Fixed     Healthy      OK                      87.4 GB 255.51 GB
```

We should clearly be using D:\ on github.  There is plenty of space and it is intended as temporary storage.
Do we have to worry about someone using this action somewhere else?

gstrauss commented:
Changing the default to D:\ for `work-dir` and the cygwin-install-action CI now takes 1 min 6 sec instead of 1 min 33 sec.  For lighttpd CI, using the D:\ drive *saves* about 2 1/2 min to install to D:\cygwin and then compile and test lighttpd (down from 10 min to about 7 1/2 min total).

gstrauss commented:
I added a test for `D:\` and fall back to `C:\` if `D:\` does not exist.

@jon-turney I believe I have addressed all of your concerns except for disk space checks on D:\ and "bullet-proofing", as you say, user-input sanity checks.  As maintainer, you're welcome to edit this PR with any tweaks that you would like.

This PR should make default (without specifying `install-dir`) use of cygwin-install-action noticeably faster.
#26 should make runs more reliable.  I'm interested in getting both of these merged.

jon-turney commented:
> Do we have to worry about someone using this action somewhere else?

No, that wasn't what I was trying to say.  Just that GitHub could change the specification and configuration of the VM whenever they feel like it.

jon-turney commented:
I've applied the work-vol changes, and added some simple tests of that option.

(When writing the tests, I was kind of a confused how the tests passed before, without the C: -> D: symlink added, since C: is hardcoded in lots of places there)

I kept the powershell -> pwsh  patch separate out of some vague concern to minimize the amount of changes.

List for me of things to look at later:
- [ ] use $true/$false rather than literals
- [ ] powershell -> pwsh  
- [ ] use ternary assignment operator where possible

gstrauss commented:
Thanks!

Now that you mention it, I thought the ternary operator required pwsh.

> use $true/$false rather than literals

I'm relatively new to pwsh, too, so please do check/verify my syntax.  As I said, I am more than happy to take suggestions.

> (When writing the tests, I was kind of a confused how the tests passed before, without the C: -> D: symlink added, since C: is hardcoded in lots of places there)

The cygwin-install-action CI did not work when `work-dir` was set until I added the C:\cygwin -> D:\cygwin symlink, so that was a good before and after test.  What did work was my lighttpd CI for cygwin, since when I specified `work-dir`, I also changed the path used when invoking the shell.  Yes, I did do some testing. :smile:

I'll rebase the PR on top of your changes shortly.

lazka commented:
> No, that wasn't what I was trying to say. Just that GitHub could change the specification and configuration of the VM whenever they feel like it.

One heuristic could be to default to "/" aka the current drive root. Each action step has "$GITHUB_WORKSPACE" set as the working directory, and it's likely that this is on a fast drive.

cclauss commented:
Please fix the git conflicts.
